### PR TITLE
add support for visa mobile (on-site)

### PR DESCRIPTION
**Pull request by arti0090**

### CODE REVIEW

**Comment by arti0090** [comment]
Input is simple for now

image

**Comment by arti0090** [comment]
image

Looks average but it works 🎉

File: tests/Api/Shop/PayingForOrdersByVisaMobileTest.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            ]),
+        );
+
+        $response = $this->client->getResponse();
+
+        $this->assertResponseCode($response, Response::HTTP_OK);
+        $this->assertResponse($response, 'shop/paying_for_orders_by_visa_mobile/test_paying_with_visa_mobile_payment_type');
+    }
+
+    private function doPlaceOrder(
```
Comment (lchrusciel) [inline tests/Api/Shop/PayingForOrdersByVisaMobileTest.php:55]: Why can't we change the definition of such a method in OrderPlacerTrait to protected?

File: tests/Unit/Api/Command/PayByVisaMobileHandlerTest.php [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    public function test_it_throws_an_exception_if_a_gateway_name_cannot_be_determined(): void
+    {
+        $this->expectException(\InvalidArgumentException::class);
+        $this->expectExceptionMessage('Gateway name cannot be determined.');
+
+        $payment = $this->prophesize(PaymentInterface::class);
+        $payment->getDetails()->willReturn([]);
+        $payment->setDetails(Argument::any());
```
Comment (lchrusciel) [inline tests/Unit/Api/Command/PayByVisaMobileHandlerTest.php:58]: is it even required?

Comment (arti0090) [reply tests/Unit/Api/Command/PayByVisaMobileHandlerTest.php:58]: 🤷 

image

**Comment by lchrusciel** [comment]
Founded bugs:
* If you use number with `+` at the beginning you are not able to fill 9 numbers of phone
* if you proceed with wrong number, you end up with 500 error 
perhaps we should add two separate boxes for phone number and skip validation entirely

**Comment by lchrusciel** [comment]
Only required change for now: let's do something with `48` prefix. Either we should support `+48` or omit this value in the first iteration.

After discussion with @jakubtobiasz we decided to hardcode 9 digits numbers and add `+48` (if needed) on backend

File: assets/shop/js/_visa_mobile.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+document.addEventListener('DOMContentLoaded', () => {
+  const COUNTRY_CODED_MOBILE_PHONE_LENGTH = 11;
+
+  let form = document.querySelector('[name="sylius_checkout_complete"]');
+  let phoneNumber = form.querySelector('[data-visa-mobile-phone-number]');
```
Comment (jakubtobiasz) [inline assets/shop/js/_visa_mobile.js:5]:
```suggestion
  let phoneNumber = form.querySelector('[data-visa-mobile-phone-number]');
  
  if (null === phoneNumber) {
  	return;
  }
```

File: config/config/sylius_fixtures.php [hunk 4]
```diff
@@ -159,6 +159,23 @@
                     ],
                     'enabled' => true,
                 ],
+                'visa_mobile' => [
+                    'code' => 'visa_mobile',
+                    'name' => 'Visa mobile (Tpay)',
+                    'gatewayFactory' => 'tpay',
+                    'gatewayName' => 'tpay',
+                    'gatewayConfig' => [
```
Comment (jakubtobiasz) [inline config/config/sylius_fixtures.php:167]: rebase and adjust setting the `gatewayConfig` accordingly to @lchrusciel changes; it's a simple thing

File: tests/Unit/Api/Command/PayByGooglePayHandlerTest.php [hunk 5]
```diff
@@ -92,6 +92,8 @@ public function test_it_throws_an_exception_if_a_payment_status_is_null(): void
                 'success_url' => null,
                 'failure_url' => null,
                 'tpay_channel_id' => null,
+                'visa_mobile_phone_number' => null,
+                'visa_mobile_phone_number' => null,
```
Comment (jakubtobiasz) [inline tests/Unit/Api/Command/PayByGooglePayHandlerTest.php:96]:
```suggestion
                'visa_mobile_phone_number' => null,
```

**Review by jakubtobiasz — COMMENTED** [review]
It'd be nice to have an E2E test here, as with VisaMobile it should be possible; check the BLIK E2E for a reference or just catch me on Slack

File: tests/Api/Shop/PayingForOrdersByVisaMobileTest.php (hunk 1 above)
Comment (lchrusciel) [reply tests/Api/Shop/PayingForOrdersByVisaMobileTest.php:135]: So why do we duplicate its content here?

File: assets/shop/js/_visa_mobile.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
+document.addEventListener('DOMContentLoaded', () => {
+  const COUNTRY_CODED_MOBILE_PHONE_LENGTH = 11;
```
Comment (lchrusciel) [inline assets/shop/js/_visa_mobile.js:2]: after our changes should this const be still at 11?

File: src/Form/Type/TpayPaymentDetailsType.php [hunk 7]
```diff
@@ -64,7 +67,28 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
...
+                        new Regex(
+                            '/^\d+$/',
+                            message: 'commerce_weavers_sylius_tpay.shop.pay.visa_mobile.regex',
+                            groups: ['sylius_checkout_complete'],
+                        ),
+                    ],
+                ],
+            );
+
+        $builder->get('visa_mobile_phone_number')->addModelTransformer(new VisaMobilePhoneDataTransformer());
```
Comment (lchrusciel) [inline src/Form/Type/TpayPaymentDetailsType.php:91]: Transformer could be injected to form type, as the event listener is

File: src/Model/PaymentDetails.php [hunk 8]
```diff
@@ -24,6 +24,7 @@ public function __construct(
         private ?string $successUrl = null,
         private ?string $failureUrl = null,
         private ?string $tpayChannelId = null,
+        private ?string $visaMobilePhoneNumber = null,
```
Comment (lchrusciel) [inline src/Model/PaymentDetails.php:27]:
```suggestion
        #[\SensitiveParameter]
        private ?string $visaMobilePhoneNumber = null,
```

**Comment by lchrusciel** [comment]
The validation do not work as expected. I allow phone numbers to be 7 digits

https://github.com/user-attachments/assets/b7933684-27b9-49e4-9d53-d0abdae54f6f

**Comment by lchrusciel** [comment]
according to their docs, the phone number `12345678904` should not be accepted, but we cannot check it, as only 9 digits are accepted in our field. Let's make it at least 9 without upper limit

**Review by lchrusciel — APPROVED** [review]

**Pull request merged**